### DISCUSSION TITLE
dwl中运行fcitx5后，alacritty里输入shift键失效，konsole没问题

### DISCUSSION THREAD

**Discussion by guyuming76** [fcitx/fcitx5#481] [Q&A]
参照 (https://github.com/djpohly/dwl/pull/12) ， 我给dwl 加了拼音输入功能：

(https://gitee.com/guyuming76/dwl/commits/guyuming1)

问题是konsole 里可以用fcitx5 输入拼音，但在 alacritty, 或 foot 里，shift 键都失效了，

这是我对问题的描述和自己的部分调试：( https://gitee.com/guyuming76/personal/commit/083f7807d05a2c958af51b42d16ce9ec355623b6 )

我不知道这是我对 dwl 的改动不对，还是 fcitx5 或 alacritty, 或 konsole 的问题，恳请高手指点方向。

更新：刚对我的dwl改动提交了一个更新，这下alacritty里shift正常了，原来keypressmod 里也要判断是否有kb_grab. 但目前 alacritty 里中文选词窗口还不显示，虽然按空格后默认中文可以出现。

**Comment by wengxt** [comment]
窗口你需要实现 input popup surface.

**Reply by guyuming76** [reply]
我把之前代码里为了能成功编译注释掉的代码都取消了注释，本地编译通过后放在了(https://gitee.com/guyuming76/dwl/tree/guyuming3)。 但这个版本我运行后在 alacritty 里按 ctrl+space 切换中文就会报 SIGSEGV 错误：

![Screenshot_20220427_212930](https://user-images.githubusercontent.com/12045548/165535893-1e97149a-ffa5-4e40-a594-c4186385884a.png)

我不确定这是不是因为 client->surface.xdg->events.unmap 里面 prev=0x51 这样的值引起的，这是不是无效的指针？ 我读到 (https://blog.csdn.net/HaoBBNuanMM/article/details/90692398) 里提到 wl_list_init 这个函数，我截图里面那些个 wl_list 会不会是没有初始化？

**Reply by guyuming76** [reply]
> 窗口你需要实现 input popup surface.

我又跟了下代码，client 变量来源于一个 wlr_surface , 即 text_input->input->focused_surface, 而 text_input->input 是 wlr_text_input_v3 类型， 我尝试的 dwl 版本引用的是 wlr_text_input_v3.h 和 wlr_input_method_v2.h ，会不会和我机器上的fcitx5 不兼容？

**Reply by wengxt** [reply]
不要说和「fcitx5」兼不兼容的问题，你compoositor 的内存问题不可能是 fcitx *直接*导致的，只有可能是你的实现不对
哪里不对那只能你自己debug了，我没有建议。你可以尝试 valgrind 帮助你调试

**Reply by guyuming76** [reply]
多谢，的确问题好像是那个 (https://github.com/djpohly/dwl/pull/12) pull request 内部代码的问题，我把我的疑问更新到那个pull request 的 conversation 里面了。

我的理解valgrind 通过一个虚拟的处理器，可以帮我们反编译一些没有源码的应用，但这个问题 wlroots, dwl, fcitx5 的源码都有，都是源码编译运行的， valgrind 能帮我啥呢？

问题是, wlroot, dwl 我都不熟，即使找出错，一时半会而让我改也不现实。

**Reply by wengxt** [reply]
可以跟踪检查内存错误，辅助你 debug

当然，如果你不想改，没能力改，也可换别的 compositor

**Reply by guyuming76** [reply]
今天运气不错，居然让我改得能运行了： https://gitee.com/guyuming76/dwl/commits/guyuming4 

![image](https://user-images.githubusercontent.com/12045548/166083344-ebb1b9f4-7cb2-4589-b3af-ece95dd774c1.png)


不过前面提到的那两个wlr 头文件引用， 确实和fcitx5里面不一样，难道是兼容的？

我也只是恰巧把别人代码调通了，里面到底咋回事，还搞不清楚，今后还请多指教，谢谢！

**Reply by wengxt** [reply]
wlr  是 wlroots 的头文件。fcitx 又不用 wlroots

**Reply by guyuming76** [reply]
我的意思是 wlr_text_input_v3.h 和 wlr_input_method_v2.h  ， 去掉 wlr 前缀应该是wayland 下的协议吧. 在fcitx5 git clone 的目录里, grep -r text_input 及 grep -r input_method ，是能找到不少内容的，不少带 zwp 前缀， 但后面 v1 啥的后缀和我手头dwl 内引用的不同
